tests/e2e/koperator_suite_test.go

```diff
@@ -17,54 +17,188 @@
+		return nil, err
+	}
+
+	classifier := tests.NewClassifier(k8sClusterPool, alltestCase)
+
+	var testPool tests.TestPool
+	testStrategy := viper.GetString(config.Tests.TestStrategy)
+
+	switch testStrategy {
+	case config.TestStrategyMinimal:
+		testPool = classifier.Minimal()
+	case config.TestStrategyVersionComplete:
+		testPool = classifier.VersionComplete()
+	case config.TestStrategyProviderComplete:
+		testPool = classifier.ProviderComplete()
+	case config.TestStrategyComplete:
+		testPool = classifier.Complete()
+	}
+
+	return testPool, nil
+}
+
 func TestKoperator(t *testing.T) {
-	RegisterFailHandler(Fail) // Note: Ginkgo - Gomega connector.
-	RunSpecs(t, "Koperator end to end test suite")
+	var err error
+
+	err = runGinkgoTests(t)
+	if err != nil {
+		t.Errorf("Koperator E2E start failed: %v", err)
+	}
 }
 
-var _ = BeforeSuite(func() {
-	By("Acquiring K8s cluster")
-	var kubeconfigPath string
-	var kubecontextName string
+func runGinkgoTests(t *testing.T) error {
+	RegisterFailHandler(Fail)
+	suiteConfig, _ := GinkgoConfiguration()
 
-	By("Acquiring K8s config and context", func() {
-		var err error
-		kubeconfigPath, kubecontextName, err = currentEnvK8sContext()
-		Expect(err).NotTo(HaveOccurred())
-	})
+	// Gomega configurations
+	format.MaxLength = 0
 
-	By("Listing kube-system pods", func() {
-		pods := k8s.ListPods(
-			ginkgo.GinkgoT(),
-			k8s.NewKubectlOptions(kubecontextName, kubeconfigPath, "kube-system"),
-			v1.ListOptions{},
-		)
+	// Run only selected tests by testID label e.g: "testID:4e980f5b5c"
+	if labelFilter := viper.GetString(config.Tests.LabelFilter); labelFilter != "" {
+		suiteConfig.LabelFilter = labelFilter
+	}
 
-		Expect(len(pods)).To(Not(BeZero()))
-	})
+	var err error
+	// Generated and load tests into the pool
+	testPool, err = beforeSuite()
+	if err != nil {
+		return fmt.Errorf("beforeSuite ran into err: %w", err)
+	}
+
+	runningSuiteProgress.allSpecCount = testPool.GetTestSuiteSpecsCount()
+
+	testSuiteDuration := testPool.GetTestSuiteDurationSerial()
+	if suiteConfig.ParallelTotal > 1 {
+		testSuiteDuration = testPool.GetTestSuiteDurationParallel()
+	}
+
+	maxTimeout, err := time.ParseDuration(viper.GetString(config.Tests.MaxTimeout))
+	if err != nil {
+		return fmt.Errorf("could not parse MaxTimeout into time.Duration: %w", err)
+	}
+	// Protection against too long test suites
+	if testSuiteDuration > maxTimeout {
+		return fmt.Errorf("tests estimated duration: '%s' bigger then maxTimeout: '%s'", testSuiteDuration.String(), maxTimeout.String())
+	}
+
+	// Calculated timeout can be overran with the specified time length
+	allowedOverrun, err := time.ParseDuration(viper.GetString(config.Tests.AllowedOverrunDuration))
+	if err != nil {
+		return fmt.Errorf("could not parse AllowedOverrunDuration into time.Duration: %w", err)
+	}
+	// Set TestSuite timeout based on the generated tests
+	suiteConfig.Timeout = testSuiteDuration + allowedOverrun
```

Comment:
Question: If we test that we don't exceed the max duration a couple of lines up from here why don't we test with this timeout value instead? If we have an overrun set we could accidentally exceed the max timeout set by the user which can cause confusion I believe.

Reply:
You have right.

My idea was that the netto test execution should not be bigger then the max timeout.
But I use for timeout the testSuiteDuration + allowedOverrun which can be bigger then the maxTimeout.

Reply:
fixed: [400c0ff](https://github.com/banzaicloud/koperator/pull/1020/commits/400c0ff9d666b916831c6613d2fc80175563f46a)